server/datastore/mysql/hosts.go

```diff
@@ -367,27 +367,42 @@ func loadHostUsersDB(ctx context.Context, db sqlx.QueryerContext, host *fleet.Ho
 }
 
 func (d *Datastore) DeleteHost(ctx context.Context, hid uint) error {
-	err := d.deleteEntity(ctx, hostsTable, hid)
-	if err != nil {
-		return ctxerr.Wrapf(ctx, err, "deleting host with id %d", hid)
+	delHostRef := func(tx sqlx.ExtContext, table string) error {
+		_, err := tx.ExecContext(ctx, fmt.Sprintf(`DELETE FROM %s WHERE host_id=?`, table), hid)
+		if err != nil {
+			return ctxerr.Wrapf(ctx, err, "deleting %s for host %d", table, hid)
+		}
+		return nil
 	}
 
-	_, err = d.writer.ExecContext(ctx, `DELETE FROM host_seen_times WHERE host_id=?`, hid)
-	if err != nil {
-		return ctxerr.Wrap(ctx, err, "deleting host seen times")
-	}
+	return d.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		_, err := tx.ExecContext(ctx, `DELETE FROM hosts WHERE id = ?`, hid)
```

Comment:
By removing the `hosts` entry first we have the risk of leaving some undeleted entries forever (e.g. say for some reason `host_seen_times` deletion fails).

By removing `hosts` entry last the user could still retry the deletion later (in case it was a temporary issue) - though there's the risk for a host with invalid/inconsistent data in that case.

Reply:
Nope, because it's all in a tx.

Reply:
Nvm, this is indeed a transaction :)

Reply:
I probably got confused between PRs.